nexus/src/app/background/tasks/support_bundle_collector.rs

```diff
@@ -513,40 +513,81 @@ impl BundleCollection {
+        // but in the meantime, we spawn this cancellation check in an entirely
+        // new tokio task. Because of this separation, each task (the one
+        // checking for cancellation, and the main thread attempting to collect
+        // the bundle) do not risk preventing the other from being polled
+        // indefinitely.
+        let mut check_for_cancellation = tokio::task::spawn({
+            let s = self.clone();
+            async move {
+                loop {
+                    // Timer fired mid-collection - let's check if we should stop.
+                    yield_interval.tick().await;
                     trace!(
-                        &self.log,
+                        &s.log,
                         "Checking if Bundle Collection cancelled";
-                        "bundle" => %self.bundle.id
+                        "bundle" => %s.bundle.id
                     );
 
-                    let bundle = self.datastore.support_bundle_get(
-                        &self.opctx,
-                        self.bundle.id.into()
-                    ).await?;
+                    let bundle = s
+                        .datastore
+                        .support_bundle_get(&s.opctx, s.bundle.id.into())
+                        .await?;
                     if !matches!(bundle.state, SupportBundleState::Collecting) {
-                        warn!(
-                            &self.log,
-                            "Support Bundle cancelled - stopping collection";
-                            "bundle" => %self.bundle.id,
-                            "state" => ?self.bundle.state
-                        );
                         anyhow::bail!("Support Bundle Cancelled");
                     }
-                },
-                // Otherwise, keep making progress on the collection itself.
-                report = &mut collection => {
-                    info!(
-                        &self.log,
-                        "Bundle Collection completed";
-                        "bundle" => %self.bundle.id
-                    );
-                    return report;
-                },
+                }
             }
+        });
+
+        tokio::select! {
+            // Returns if the bundle has been cancelled explicitly, or if we
+            // cannot successfully check the bundle state.
+            why = &mut check_for_cancellation => {
+                let why = why.expect("Should not cancel the bundle-checking task without returning");
+                warn!(
+                    &self.log,
+                    "Support Bundle cancelled - stopping collection";
+                    "bundle" => %self.bundle.id,
+                    "state" => ?self.bundle.state
+                );
+                return why;
+            },
+            // Otherwise, keep making progress on the collection itself.
+            report = &mut collection => {
```

Comment:
Nit: since there's now only one `select!`, rather than a `select!` in a loop, we don't _need_ the `&mut` here (though we _do_ need the `&mut check_for_cancellation` in the other select arm, so that we can cancel it here, but using `AbortOnDropHandle` would obviate that...)
```suggestion
            report = collection => {
```